slant.1
.\"	$Id$
.\"
.\" Copyright (c) 2018 Kristaps Dzonsons <kristaps@bsd.lv>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt SLANT 1
.Os
.Sh NAME
.Nm slant
.Nd client for remote system monitoring
.Sh SYNOPSIS
.Nm slant
.Op Fl f Ar config
.Op Fl o Ar order
.Op Ar url...
.Sh DESCRIPTION
The
.Nm
utility displays system statistics gathered from resources listed in
.Pa ~/.slantrc .
These resource must be running
.Xr slant-cgi 8 .
Its arguments are as follows:
.Bl -tag -width Ds
.It Fl f Ar config
Specify an alternate configuration location.
.It Fl o Ar order
Default order of host listing.
May be
.Ar cmdline
for natural order,
.Ar host
for hostname,
.Ar cpu
for immediate CPU, or
.Ar mem
for immediate memory.
.It Ar url
Override the configuration's hosts with those provided.
See
.Sx URLs
for details.
.El
.Pp
By default, hosts are ordered as given in the configuration and are
queried every 60 seconds.
If hosts are passed as arguments to
.Nm ,
they are used instead of the configuration file's.
.Pp
If not overridden in the configuration, host status is displayed as
if given the following configuration:
.Bd -literal -offset indent
layout {
  header ;
  errlog 10 ;
  host {
    cpu qmin_bars qmin hour ;
    mem qmin_bars qmin hour ;
    nprocs qmin_bars qmin hour ;
    net qmin hour ;
    disc qmin hour ;
    link ip state access ;
    rprocs qmin ;
    host record ;
  } ;
} ;
.Ed
.Pp
.Nm
needs a minimum of about 90 columns to display data, plus more depending
upon the length of the domain names.
If the window isn't wide enough, the following are elided until the
minimum size is reached:
.Pp
.Bl -enum -compact
.It
bar charts
.It
hourly records
.It
link IP and state
.El
.Pp
Data is not recorded to the local machine: it is
.Qq live .
So if the front-end is shut down and restarted with hosts that are
unresponsive, their data will not be refreshed til the next access.
.Pp
An error and debug log is shown below the table of all hosts.
The log is saved to
.Pa ~/.slant-errlog .
.Ss Configuration file
The configuration file, which defaults to
.Pa ~/.slantrc ,
is structured with the following top-level nodes.
All tokens are white-space separated, so
.Qq servers foo;
species a server named
.Qq foo; .
Literals are case sensitive.
.Bd -literal -offset indent
"waittime" NUM ";"
"timeout" NUM ";"
.Ed
.Pp
The global
.Li waittime
before each host is processed.
The countdown for each host begins after its last disconnect.
The minimum is 15 secons.
Also the global
.Li timeout
after which data transfers are terminated (following the last byte
transferred).
The minimum is 10 seconds.
.Bd -literal -offset indent
"servers" url [url...] ["{"
  ["waittime" NUM ";" ]
  ["timeout" NUM ";" ]
"}"] ";"
.Ed
.Pp
There may be several
.Cm servers
sections, each with different server arguments overriding the globals.
See
.Sx URLs
for details on the
.Li url
field.
.Bd -literal -offset indent
"layout" "{"
  ["header" ";"]
  ["host" "{" [column]+ "}" ";"
  ["errlog" NUM ";"]
"}" ";"
.Ed
.Pp
The
.Cm layout
block configures how host are displayed.
If
.Cm header
is specified, a column header is shown at the top of the screen.
If
.Cm errlog
is non-zero, it is the rows in the error/debug window.
The
.Cm host
block consists of a series of column types and arguments.
Each column may have up to six lines of content.
If the column arguments do not have a line specification, they're
assumed to be on the first line.
In the event of duplicate lines, only the last is retained.
A column for a host ends in a semicolon or the close-brace for the host
layout block.
.Bd -literal -offset indent
"host" "{"
 [ coltype
   [ "line1" "{" colargs "}" |
     "line2" "{" colargs "}" |
     "line3" "{" colargs "}" |
     "line4" "{" colargs "}" |
     "line5" "{" colargs "}" |
     "line6" "{" colargs "}" |
     colargs
   ] ";" ]+
"}" ";"]
.Ed
.Pp
The column types and arguments are as follows:
.Bd -literal -offset indent
"cpu" [time_interval_bars|time_interval]+
"mem" [time_interval_bars|time_interval]+
"net" [time_interval]+
"disc" [time_interval]+
"link" ["ip"|"state"|"access"]+
"host" ["record"|"slant_version"|"uptime"|"clock_drift"|uname]+
"nprocs" [time_interval_bars|time_interval]+
"rprocs" [time_interval_bars|time_interval]+
"nfiles" [time_interval_bars|time_interval]+
.Ed
.Pp
The
.Cm time_interval_bars
fields draw coloured bar graph when specifying
.Cm qmin_bars ,
for the quarter-minute summary;
.Cm min_bars ,
the minute summary;
.Cm hour_bars ,
the hourly summary;
.Cm day_bars ,
the daily summary;
.Cm week_bars ,
the weekly summary; and
.Cm year_bars ,
the yearly summary.
The
.Cm time_interval
writes a percentage when specifying
.Cm qmin ,
for the quarter-minute summary;
.Cm min ,
for the minute summary;
.Cm hour ,
for the hourly summary;
.Cm day ,
for the daily summary;
.Cm week ,
for the weekly summary; or
.Cm year ,
for the yearly summary.
These summaries depend upon the category.
.Pp
The
.Cm uname
fields may be
.Cm machine ,
for host hardware;
.Cm osversion ,
for OS-specific version;
.Cm osrelease ,
for OS-specific release; and
.Cm osname ,
for operating system name.
.Bl -tag -width Ds
.It Cm cpu
CPU time averaged across all cores/CPUs on the system.
Summaries are shown as percentages.
Percentages more than 80% are coloured red; more than 50%, yellow.
The bar graph of the instantaneous view coloured in the same way.
.It Cm mem
Memory usage (all active memory over all pages).
Summaries are shown as percentage.
Percentages more than 80% are coloured red; more than 50%, yellow.
The bar graph of the instantaneous view coloured in the same way.
.It Cm net
Data received and transmitted as averaged over all network devices.
Summaries are in human-readable scaled units (e.g., KB/s).
.It Cm disc
Data read and written as averaged over all configured devices.
This is shown as an average over two intervals: in the last 15
Summaries are in human-readable scaled units (e.g., KB/s).
.It Cm link
If requesting
.Cm ip ,
the IP address (IPV4 or IPV6) of the host and the connection state.
If
.Cm state ,
may be one of
.Li strt ,
startup;
.Li rslv ,
resolving;
.Li idle ,
waiting for next connection;
.Li cnrd ,
ready to connect;
.Li cnct ,
connecting;
.Li cldn ,
connection finished (success);
.Li cler ,
connection finished (error);
.Li wrte ,
writing request; or
.Li read ,
reading response.
Lastly,
.Cm access
is the time since last ping.
Shown as hours, minutes, seconds elapsed.
If a worrying amount of elapsed time has shown, the time will be shown
in yellow.
If the amount indicates problems, it will be shown in red.
.It Cm host
If
.Cm record ,
the last data collection time as recorded by the remote host's
collection system.
Shown as hours, minutes, seconds elapsed.
If a worrying amount of elapsed time has shown, the time will be shown
in yellow.
If the amount indicates problems, it will be shown in red.
The last connection time as recorded by the local host's
Shown as hours, minutes, seconds elapsed.
If a worrying amount of elapsed time has shown, the time will be shown
in yellow.
If the amount indicates problems, it will be shown in red.
The
.Cm slant_version ,
column argument prints the collector version.
Lastly,
.Cm uptime
prints the uptime in days, hours, and minutes.
.It Cm nprocs
The number of running processes over the maximum configured amount.
Summaries are in percentages.
Percentages more than 80% are coloured red; more than 50%, yellow.
The bar graph of the instantaneous view is coloured in the same way.
.It Cm rprocs
Of the commands given to the collector for monitoring, the percentage
that are running.
Summaries are in percentages.
If no commands were given, is always 100%.
If less than 100%, shown in red.
.It Cm nfiles
The number of open files over the maximum possible amount.
Summaries are in percentages.
Percentages more than 80% are coloured red; more than 50%, yellow.
The bar graph of the instantaneous view is coloured in the same way.
.El
.Pp
The hostname (domain name) is always shown first.
.Ss URLs
The URLs passed on the command line or in the configuration file are in
the following format:
.Bd -literal
[https://|http://][username:password@]host[:port][/path][?query]
.Ed
.Pp
If the schema is not provided, HTTP is used regardless of the port (if
given).
The username and password should only be used with HTTPS, but this is
not mandated.
.\" The following requests should be uncommented and used where appropriate.
.\" .Sh CONTEXT
.\" For section 9 functions only.
.\" .Sh RETURN VALUES
.\" For sections 2, 3, and 9 function return values only.
.\" .Sh ENVIRONMENT
.\" For sections 1, 6, 7, and 8 only.
.\" .Sh FILES
.\" .Sh EXIT STATUS
.\" For sections 1, 6, and 8 only.
.Sh EXAMPLES
To query the localhost and a remote machine:
.Bd -literal
% slant -o host
.Ed
.Pp
With the following configuration file:
.Bd -literal -offset indent
waittime 60 ;
servers
  https://remote/slant-cgi
  ;
servers
  localhost/cgi-bin/slant-cgi
  { waittime 15 } ;
.Ed
.Pp
This establishes a default wait time of 60 seconds.
This value is used by the remote host.
The local host has its wait time overriden at 15 seconds.
It uses the default node display.
.Pp
The following is optimised for a quick look at memory and CPU usage of
many hosts, so the error log and header are suppressed.
.Bd -literal -offset indent
waittime 60 ;
servers
  https://remote1/slant-cgi
  https://remote2/slant-cgi
  https://remote3/slant-cgi
  ;
servers
  localhost/cgi-bin/slant-cgi
  { waittime 15 } ;
layout {
  host {
    cpu qmin_bars qmin min hour day ;
    mem qmin_bars qmin min hour day ;
    link access ;
  }
} ;
.Ed
.Pp
A more vertical layout for the same is as follows.
This stacks information using the multi-line features.
.Bd -literal -offset indent
layout {
  host {
    cpu
      line1 { qmin_bars qmin }
      line2 { min_bars min }
      line3 { hour_bars hour }
      line4 { day_bars day }
      line5 { week_bars week }
      line6 { year_bars year } ;
    mem
      line1 { qmin_bars qmin }
      line2 { min_bars min }
      line3 { hour_bars hour }
      line4 { day_bars day }
      line5 { week_bars week }
      line6 { year_bars year } ;
    link
      line1 { access state }
      line2 { ip } ;
  }
} ;
.Ed
.\" .Sh DIAGNOSTICS
.\" For sections 1, 4, 6, 7, 8, and 9 printf/stderr messages only.
.\" .Sh ERRORS
.\" For sections 2, 3, 4, and 9 errno settings only.
.Sh SEE ALSO
.Xr slant-collectd 8
.\" .Sh STANDARDS
.\" .Sh HISTORY
.\" .Sh AUTHORS
.\" .Sh CAVEATS
.\" .Sh BUGS


slant.c
/*	$Id$ */
/*
 * Copyright (c) 2018 Kristaps Dzonsons <kristaps@bsd.lv>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include "config.h"

#if HAVE_SYS_QUEUE
# include <sys/queue.h>
#endif
#include <sys/poll.h>
#include <sys/socket.h>
#include <arpa/inet.h>

#include <assert.h>
#include <curses.h>
#if HAVE_ERR
# include <err.h>
#endif
#include <errno.h>
#include <limits.h>
#include <locale.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <tls.h>
#include <unistd.h>

#include <kcgi.h>
#include <kcgijson.h>

#include "extern.h"
#include "slant.h"
#include "json.h"

static	volatile sig_atomic_t sigged;

static void
dosig(int code)
{

	sigged = 1;
}

void
recset_free(struct recset *r)
{

	if (NULL == r)
		return;

	free(r->version);
	jsmn_system_clear(&r->system);
	jsmn_record_free_array(r->byqmin, r->byqminsz);
	jsmn_record_free_array(r->bymin, r->byminsz);
	jsmn_record_free_array(r->byhour, r->byhoursz);
	jsmn_record_free_array(r->byday, r->bydaysz);
	jsmn_record_free_array(r->byweek, r->byweeksz);
	jsmn_record_free_array(r->byyear, r->byyearsz);
}

static void
nodes_free(struct node *n, size_t sz)
{
	size_t	 i;

	for (i = 0; i < sz; i++) {
		if (NULL != n[i].xfer.tls)
			tls_close(n[i].xfer.tls);
		if (-1 != n[i].xfer.pfd->fd)
			close(n[i].xfer.pfd->fd);
		tls_free(n[i].xfer.tls);
		free(n[i].host);
		free(n[i].xfer.wbuf);
		free(n[i].xfer.rbuf);
		recset_free(n[i].recs);
		free(n[i].recs);
		free(n[i].httpauth);
	}

	free(n);
}

/*
 * Run a single step of the state machine.
 * For each node, we examine the current state (n->state) and perform
 * some task based upon that.
 * Return <0 on some sort of error condition or the number of nodes that
 * have changed, which means some sort of window update event should
 * occur to reflect the change.
 */
static int
nodes_update(struct out *out, struct node *n, size_t sz, time_t t)
{
	size_t	 i;
	int	 dirty = 0;

	for (i = 0; i < sz; i++) {
		switch (n[i].state) {
		case STATE_CONNECT_WAITING:
			if (n[i].waitstart + n[i].waittime >= t) 
				break;
			n[i].state = STATE_CONNECT_READY;
			n[i].dirty = 1;
			break;
		case STATE_CONNECT_READY:
			if ( ! http_init_connect(out, &n[i], t))
				return -1;
			break;
		case STATE_CONNECT:
			if ( ! http_connect(out, &n[i], t))
				return -1;
			break;
		case STATE_WRITE:
			if ( ! http_write(out, &n[i], t))
				return -1;
			break;
		case STATE_CLOSE_ERR:
			if ( ! http_close_err(out, &n[i], t))
				return -1;
			break;
		case STATE_CLOSE_DONE:
			if ( ! http_close_done(out, &n[i], t))
				return -1;
			break;
		case STATE_READ:
			if ( ! http_read(out, &n[i], t))
				return -1;
			break;
		default:
			abort();
		}

		if (n[i].dirty)
			dirty++;
	}

	return dirty;
}

/*
 * Sort comparator for memory usage.
 * Needs to be run once per iteration.
 */
static int
cmp_mem(const void *p1, const void *p2)
{
	const struct node *n1 = p1, *n2 = p2;

	if (NULL == n1->recs ||
	    0 == n1->recs->byqminsz)
		return 1;
	if (NULL == n2->recs ||
	    0 == n2->recs->byqminsz)
		return -1;
	if (n1->recs->byqmin[0].mem <
	    n2->recs->byqmin[0].mem)
		return 1;
	if (n1->recs->byqmin[0].mem >
	    n2->recs->byqmin[0].mem)
		return -1;
	return 0;
}

/*
 * Sort comparator for CPU time.
 * Needs to be run once per iteration.
 */
static int
cmp_cpu(const void *p1, const void *p2)
{
	const struct node *n1 = p1, *n2 = p2;

	if (NULL == n1->recs ||
	    0 == n1->recs->byqminsz)
		return 1;
	if (NULL == n2->recs ||
	    0 == n2->recs->byqminsz)
		return -1;
	if (n1->recs->byqmin[0].cpu <
	    n2->recs->byqmin[0].cpu)
		return 1;
	if (n1->recs->byqmin[0].cpu >
	    n2->recs->byqmin[0].cpu)
		return -1;
	return 0;
}

/*
 * Sort comparator for hostnames.
 * Needs to only be run once for the list.
 */
static int
cmp_host(const void *p1, const void *p2)
{
	const struct node *n1 = p1, *n2 = p2;

	return strcmp(n1->host, n2->host);
}

/*
 * Common leading material for all logging messages.
 */
static void
xloghead(struct out *out)
{
	char	 	 buf[32];
	struct tm	*tm;
	time_t		 t = time(NULL);

	assert(NULL != out->errwin);
	tm = localtime(&t);
	strftime(buf, sizeof(buf), "%F %T", tm);
	waddstr(out->errwin, buf);
	fprintf(out->errs, "%s: ", buf);
	wprintw(out->errwin, " %lc ", L'\x2502');
}

/*
 * Emit warning message with errno.
 */
void
xwarn(struct out *out, const char *fmt, ...)
{
	va_list  ap;
	int	 er = errno;

	if (NULL != out->errwin) {
		xloghead(out);
		if (NULL != fmt) {
			va_start(ap, fmt);
			vwprintw(out->errwin, fmt, ap);
			va_end(ap);
		}
		wprintw(out->errwin, "%s%s (%d)\n", 
			NULL == fmt ? "" : ": ", strerror(er), er);
		wrefresh(out->errwin);
	}

	if (NULL != fmt) {
		va_start(ap, fmt);
		vfprintf(out->errs, fmt, ap);
		va_end(ap);
	}
	fprintf(out->errs, "%s%s (%d)\n", 
		NULL == fmt ? "" : ": ", strerror(er), er);
	fflush(out->errs);
}

/*
 * Emit warning message.
 */
void
xwarnx(struct out *out, const char *fmt, ...)
{
	va_list ap;

	if (NULL != out->errwin) {
		xloghead(out);
		wattron(out->errwin, A_BOLD);
		waddstr(out->errwin, "Warning");
		wattroff(out->errwin, A_BOLD);
		waddstr(out->errwin, ": ");
		if (NULL != fmt) {
			va_start(ap, fmt);
			vwprintw(out->errwin, fmt, ap);
			va_end(ap);
		}
		waddch(out->errwin, '\n');
		wrefresh(out->errwin);
	}

	fprintf(out->errs, "Warning: ");
	va_start(ap, fmt);
	vfprintf(out->errs, fmt, ap);
	va_end(ap);
	fputc('\n', out->errs);
	fflush(out->errs);
}

/*
 * Emit debugging message if "out->debug" has been set.
 */
void
xdbg(struct out *out, const char *fmt, ...)
{
	va_list ap;

	if ( ! out->debug) 
		return;

	if (NULL != out->errwin) {
		xloghead(out);
		if (NULL != fmt) {
			va_start(ap, fmt);
			vwprintw(out->errwin, fmt, ap);
			va_end(ap);
		}
		waddch(out->errwin, '\n');
		wrefresh(out->errwin);
	}

	if (NULL != fmt) {
		va_start(ap, fmt);
		vfprintf(out->errs, fmt, ap);
		va_end(ap);
	}
	fputc('\n', out->errs);
	fflush(out->errs);
}

/*
 * Construct a default layout depending on "maxx", the maximum number of
 * available columns, "maxy" for rows, and "n" nodes of length "nsz".
 * Put the results in "d".
 * This copies from the configuration, if applicable, or sets defaults.
 * Return zero on failure (fatal), non-zero on success.
 */
static int
layout(struct config *cfg, struct out *out, size_t maxx, 
	size_t maxy, const struct node *n, size_t nsz, struct draw *d)
{
	size_t	 i;

	if (NULL != cfg->draw) {
		d->header = cfg->draw->header;
		d->errlog = cfg->draw->errlog;
		if (d->errlog >= maxy - d->header) 
			return 0;
	} else {
		d->header = 1;
		if (maxy > 60)
			d->errlog = 10;
		else if (maxy > 40)
			d->errlog = 5;
	}

#define	ORD_CPU 	0 /* Default order... */
#define	ORD_MEM		1
#define	ORD_PROCS	2
#define	ORD_NET		3
#define	ORD_DISC	4
#define	ORD_LINK	5
#define	ORD_RPROCS	6
#define	ORD_HOST	7

	if (NULL != cfg->draw && cfg->draw->boxsz) {
		d->boxsz = cfg->draw->boxsz;
		d->maxline = cfg->draw->maxline;
		d->box = calloc(d->boxsz, sizeof(struct drawbox));
		if (NULL == d->box)
			return -1;
		for (i = 0; i < d->boxsz; i++)
			d->box[i] = cfg->draw->box[i];
	} else {
		d->boxsz = 8;
		d->maxline = 1;
		d->box = calloc(d->boxsz, sizeof(struct drawbox));
		if (NULL == d->box)
			return -1;

		d->box[ORD_CPU].cat = DRAWCAT_CPU;
		d->box[ORD_MEM].cat = DRAWCAT_MEM;
		d->box[ORD_NET].cat = DRAWCAT_NET;
		d->box[ORD_DISC].cat = DRAWCAT_DISC;
		d->box[ORD_PROCS].cat = DRAWCAT_PROCS;
		d->box[ORD_LINK].cat = DRAWCAT_LINK;
		d->box[ORD_RPROCS].cat = DRAWCAT_RPROCS;
		d->box[ORD_HOST].cat = DRAWCAT_HOST;

		d->box[ORD_CPU].lines[0].line = 
			LINE_QMIN_BARS | LINE_QMIN | LINE_HOUR;
		d->box[ORD_MEM].lines[0].line = 
			LINE_QMIN_BARS | LINE_QMIN | LINE_HOUR;
		d->box[ORD_NET].lines[0].line = 
			LINE_QMIN | LINE_HOUR;
		d->box[ORD_DISC].lines[0].line = 
			LINE_QMIN | LINE_HOUR;
		d->box[ORD_PROCS].lines[0].line = 
			LINE_QMIN_BARS | LINE_QMIN | LINE_HOUR;
		d->box[ORD_LINK].lines[0].line = 
			LINK_IP | LINK_STATE | LINK_ACCESS;
		d->box[ORD_RPROCS].lines[0].line = LINE_QMIN;
		d->box[ORD_HOST].lines[0].line = HOST_RECORD;

		if (maxx > compute_width(n, nsz, d)) 
			return 1;

		d->box[ORD_CPU].lines[0].line &= ~LINE_QMIN_BARS;
		d->box[ORD_MEM].lines[0].line &= ~LINE_QMIN_BARS;
		d->box[ORD_PROCS].lines[0].line &= ~LINE_QMIN_BARS;

		if (maxx > compute_width(n, nsz, d)) 
			return 1;

		d->box[ORD_CPU].lines[0].line &= ~LINE_HOUR;
		d->box[ORD_MEM].lines[0].line &= ~LINE_HOUR;
		d->box[ORD_NET].lines[0].line &= ~LINE_HOUR;
		d->box[ORD_DISC].lines[0].line &= ~LINE_HOUR;
		d->box[ORD_PROCS].lines[0].line &= ~LINE_HOUR;

		if (maxx > compute_width(n, nsz, d)) 
			return 1;

		d->box[ORD_LINK].lines[0].line &= ~(LINK_IP | LINK_STATE);
	}

	return maxx > compute_width(n, nsz, d);
}

int
main(int argc, char *argv[])
{
	int	 	 c, first = 1, maxy, maxx;
	size_t		 i, sz;
	const char	*cfgfile = NULL;
	struct node	*n = NULL;
	struct pollfd	*pfds = NULL;
	struct timespec	 ts;
	sigset_t	 mask, oldmask;
	time_t		 last, now;
	char		*cp;
	struct draw	 d;
	struct config	 cfg;
	struct out	 out;
	
	if (NULL == getenv("HOME")) 
		errx(EXIT_FAILURE, "no HOME directory defined");
	if (NULL == setlocale(LC_ALL, ""))
		err(EXIT_FAILURE, NULL);

	/* Initial pledge. */

#if HAVE_PLEDGE
	if (-1 == pledge("cpath wpath tty rpath dns inet stdio", NULL))
		err(EXIT_FAILURE, NULL);
#endif

	memset(&out, 0, sizeof(struct out));
	memset(&d, 0, sizeof(struct draw));
	memset(&cfg, 0, sizeof(struct config));

	out.debug = 1;

	/*
	 * Open $HOME/.slant-errlog to catch any errors.
	 * This way our error buffer in the window is saved.
	 * Do this with the "cpath wpath" cause we might create the file
	 * in doing so.
	 */

	c = asprintf(&cp, "%s/.slant-errlog", getenv("HOME"));
	if (c < 0)
		err(EXIT_FAILURE, NULL);
	if (NULL == (out.errs = fopen(cp, "a")))
		err(EXIT_FAILURE, "%s", cp);
	free(cp);

	/* Repledge by dropping the error log pledges. */

#if HAVE_PLEDGE
	if (-1 == pledge("tty rpath dns inet stdio", NULL))
		err(EXIT_FAILURE, NULL);
#endif

	/* Start up TLS handling really early. */

	if (tls_init() < 0)
		err(EXIT_FAILURE, NULL);

	/* 
	 * Establish our signal handling: have TERM, QUIT, and INT
	 * interrupt the poll and cause us to exit.
	 * Otherwise, we block the signal.
	 * TODO: handle SINGWINCH.
	 */

	if (sigemptyset(&mask) < 0)
		err(EXIT_FAILURE, NULL);
	if (SIG_ERR == signal(SIGTERM, dosig))
		err(EXIT_FAILURE, NULL);
	if (SIG_ERR == signal(SIGQUIT, dosig))
		err(EXIT_FAILURE, NULL);
	if (SIG_ERR == signal(SIGINT, dosig))
		err(EXIT_FAILURE, NULL);
	if (sigaddset(&mask, SIGTERM) < 0)
		err(EXIT_FAILURE, NULL);
	if (sigaddset(&mask, SIGQUIT) < 0)
		err(EXIT_FAILURE, NULL);
	if (sigaddset(&mask, SIGINT) < 0)
		err(EXIT_FAILURE, NULL);
	if (sigprocmask(SIG_BLOCK, &mask, &oldmask) < 0)
		err(EXIT_FAILURE, NULL);

	/* Parse arguments. */

	while (-1 != (c = getopt(argc, argv, "f:o:w:"))) 
		switch (c) {
		case 'f':
			cfgfile = strdup(optarg);
			break;
		case 'o':
			if (0 == strcmp(optarg, "host"))
				d.order = DRAWORD_HOST;
			else if (0 == strcmp(optarg, "cmdline"))
				d.order = DRAWORD_CMDLINE;
			else if (0 == strcmp(optarg, "cpu"))
				d.order = DRAWORD_CPU;
			else if (0 == strcmp(optarg, "mem"))
				d.order = DRAWORD_MEM;
			else
				goto usage;
			break;
		default:
			goto usage;
		}

	argc -= optind;
	argv += optind;

	/*
	 * Parse our configuration file.
	 * This will tell us all we need to know about our runtime.
	 * Use "cp" as a temporary variable in case we need to create
	 * the configuration file name on-demand.
	 */

	cp = NULL;
	if (NULL == cfgfile) {
		c = asprintf(&cp, "%s/.slantrc", getenv("HOME"));
		if (c < 0)
			err(EXIT_FAILURE, NULL);
		cfgfile = cp;
	} 

	c = config_parse(cfgfile, &cfg, argc, argv);
	free(cp);
	if ( ! c)
		return EXIT_FAILURE;

	/* 
	 * Initialise data.
	 * On any failure---which in this block will be memory
	 * failure---just exit the program immediately.
	 * We don't really have any state to clean up at this point.
	 */
	
	if (0 == cfg.urlsz)
		errx(EXIT_FAILURE, "no urls given");

	n = calloc(cfg.urlsz, sizeof(struct node));
	if (NULL == n)
		err(EXIT_FAILURE, NULL);

	pfds = calloc(cfg.urlsz, sizeof(struct pollfd));
	if (NULL == pfds)
		err(EXIT_FAILURE, NULL);

	for (i = 0; i < cfg.urlsz; i++) {
		pfds[i].fd = -1;
		n[i].xfer.pfd = &pfds[i];
		n[i].state = STATE_STARTUP;
		n[i].url = cfg.urls[i].url;
		n[i].waittime = 
			cfg.urls[i].waittime ?
			cfg.urls[i].waittime : (time_t)cfg.waittime;
		n[i].timeout = 
			cfg.urls[i].timeout ?
			cfg.urls[i].timeout : (time_t)cfg.timeout;
		dns_parse_url(&out, &n[i]);
	}

	/* 
	 * All data initialised.
	 * Get our window system ready to roll.
	 * Once we initialise the screen, we're going to need to use our
	 * "errs" and "errwin" to report errors, as stderr will just
	 * munge on the screen.
	 */

	if (NULL == initscr() ||
	    ERR == start_color() ||
	    ERR == cbreak() ||
	    ERR == noecho() ||
	    ERR == nonl())
		exit(EXIT_FAILURE);

	curs_set(0);
	init_pair(1, COLOR_YELLOW, COLOR_BLACK);
	init_pair(2, COLOR_RED, COLOR_BLACK);
	getmaxyx(stdscr, maxy, maxx);

	/* Configure what we see, bailing if it's not possible. */

	c = layout(&cfg, &out, maxx, maxy, n, cfg.urlsz, &d);
	if (c < 0) {
		endwin();
		warn(NULL);
		goto out;
	} else if (0 == c) {
		endwin();
		warnx("insufficient screen dimensions");
		goto out;
	}

	assert((size_t)maxy > d.errlog);
	out.mainwin = subwin(stdscr, maxy - d.errlog, maxx, 0, 0);
	if (d.errlog) {
		out.errwin = subwin(stdscr, 0, maxx, maxy - d.errlog, 0);
		scrollok(out.errwin, 1);
	}

	for (i = 0; i < cfg.urlsz; i++) {
		n[i].state = STATE_RESOLVING;
		if ( ! dns_resolve(&out, n[i].host, &n[i].addrs))
			goto out;
		n[i].state = STATE_CONNECT_READY;
	}

	/* 
	 * FIXME: rpath needed by libressl.
	 * We can relieve this by pre-loading our certs.
	 */

#if HAVE_PLEDGE
	if (-1 == pledge("tty rpath inet stdio", NULL))
		err(EXIT_FAILURE, NULL);
#endif

	/* Main loop. */

	ts.tv_sec = 1;
	ts.tv_nsec = 0;
	last = 0;

	while ( ! sigged) {
		now = time(NULL);
		if ((c = nodes_update(&out, n, cfg.urlsz, now)) < 0)
			break;

		/* Re-sort, if applicable. */

		sz = sizeof(struct node);
		switch (d.order) {
		case DRAWORD_CMDLINE:
			break;
		case DRAWORD_CPU:
			qsort(n, cfg.urlsz, sz, cmp_cpu);
			break;
		case DRAWORD_HOST:
			/* Only do this once. */
			if (0 == first)
				break;
			qsort(n, cfg.urlsz, sz, cmp_host);
			break;
		case DRAWORD_MEM:
			qsort(n, cfg.urlsz, sz, cmp_mem);
			break;
		}

		/*
		 * Update if our data is dirty (c > 0) or if we're on
		 * the first iteration, just to show something.
		 * If we've nothing to show but more than one second has
		 * passed, then simply update the time displays.
		 */

		if ((c || first) && now > last) {
			draw(&out, &d, n, cfg.urlsz, now);
			for (i = 0; i < cfg.urlsz; i++) 
				n[i].dirty = 0;
			wrefresh(out.mainwin);
			first = 0;
		} else if (now > last) {
			drawtimes(&out, &d, n, cfg.urlsz, now);
			wrefresh(out.mainwin);
		}

		last = now;
		if (ppoll(pfds, cfg.urlsz, &ts, &oldmask) < 0 && 
		    EINTR != errno) {
			xwarn(&out, "poll");
			break;
		}
	}

out:
	if ( ! isendwin()) {
		if (NULL != out.errwin)
			delwin(out.errwin);
		delwin(out.mainwin);
		endwin();
	}
	nodes_free(n, cfg.urlsz);
	config_free(&cfg);
	free(d.box);
	free(pfds);
	return EXIT_SUCCESS;
usage:
	fprintf(stderr, "usage: %s "
		"[-f conf] "
		"[-o order] "
		"[url...]\n",
		getprogname());
	return EXIT_FAILURE;
}
